src/widgets/NumberEntryWidget.zig

```diff
@@ -0,0 +1,159 @@
+    return struct {
+        wd: WidgetData = undefined,
+        text_box: dvui.TextEntryWidget = undefined,
+        init_opts: InitOptions,
+        src: std.builtin.SourceLocation,
+
+        pub const InitOptions =
+            struct {
+            buffer: []u8,
+            min: ?T = null,
+            max: ?T = null,
+
+            pub fn audit(self: InitOptions) void {
+                if (self.max != null and self.min != null) {
+                    std.debug.assert(self.max.? > self.min.?);
+                }
+            }
+        };
+
+        const base_filter = "abcdfghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ`~!@#$%^&*()_=[{]}\\|;:'\",<>/?] ";
+        const filter = switch (@typeInfo(T)) {
+            .Int => |int| switch (int.signedness) {
+                .signed => base_filter ++ ".e",
+                .unsigned => base_filter ++ "+-.e",
+            },
+            .Float => base_filter,
+            else => unreachable,
+        };
+
+        pub fn drawBackground(self: *@This()) !void {
+            try self.wd.borderAndBackground(.{});
+        }
+
+        pub fn widget(self: *@This()) Widget {
+            return Widget.init(self, data, rectFor, screenRectScale, minSizeForChild, processEvent);
+        }
+
+        pub fn data(self: *@This()) *WidgetData {
+            return &self.wd;
+        }
+
+        pub fn rectFor(self: *@This(), id: u32, min_size: Size, e: Options.Expand, g: Options.Gravity) Rect {
+            return dvui.placeIn(self.wd.contentRect().justSize(), dvui.minSize(id, min_size), e, g);
+        }
+
+        pub fn screenRectScale(self: *@This(), rect: Rect) RectScale {
+            return self.wd.contentRectScale().rectToRectScale(rect);
+        }
+
+        pub fn minSizeForChild(self: *@This(), s: Size) void {
+            self.wd.minSizeMax(self.wd.padSize(s));
+        }
+
+        pub fn processEvent(self: *@This(), e: *Event, bubbling: bool) void {
+            _ = bubbling;
+            if (e.bubbleable()) {
+                self.wd.parent.processEvent(e, true);
+            }
+        }
+
+        pub fn install(self: *@This()) !void {
+            try self.wd.register();
+            dvui.parentSet(self.widget());
+
+            if (self.wd.visible()) {
+                try dvui.tabIndexSet(self.wd.id, self.wd.options.tab_index);
+            }
+
+            try self.wd.borderAndBackground(.{});
+
+            self.text_box = TextEntryWidget.init(self.src, .{ .text = self.init_opts.buffer }, .{ .id_extra = 1 });
+            try self.text_box.install();
+
+            var buffer_backup: [256]u8 = .{0} ** 256;
+            std.mem.copyForwards(u8, &buffer_backup, self.getText());
+
+            self.text_box.processEvents();
+
+            for (std.mem.sliceTo(filter, 0), 0..) |_, i| {
+                self.text_box.filterOut(filter[i..][0..1]);
+            }
+
+            const text = self.getText();
+            const value = self.getValue();
+            if (text.len >= 2) {
```

Comment:
Are you sure `text.len` has not been modified after `self.text_box.processEvents();`?

Reply:
Good point. I'll change this